src/main/java/org/zalando/problem/spring/web/advice/AdviceTrait.java

```diff
@@ -113,25 +114,30 @@
     }
 
     default ThrowableProblem toProblem(final Throwable throwable) {
-        final StatusType status = resolveStatus(throwable)
-                .<StatusType>map(HttpStatusAdapter::new)
+        final StatusType status = Optional.ofNullable(resolveResponseStatus(throwable))
+                .map(responseStatusToStatusType())
                 .orElse(Status.INTERNAL_SERVER_ERROR);
 
         return toProblem(throwable, status);
     }
 
-    default Optional<HttpStatus> resolveStatus(final Throwable type) {
-        @Nullable final ResponseStatus candidate = findMergedAnnotation(type.getClass(), ResponseStatus.class);
-
-        if (candidate == null) {
-            if (type.getCause() == null) {
-                return Optional.empty();
+    default Function<ResponseStatus, StatusType> responseStatusToStatusType() {
+        return responseStatus -> new StatusType() {
+            @Override
+            public int getStatusCode() {
+                return responseStatus.code().value();
             }
 
-            return resolveStatus(type.getCause());
-        }
+            @Override
+            public String getReasonPhrase() {
+                return responseStatus.reason().isEmpty() ? responseStatus.value().getReasonPhrase() : responseStatus.reason();
```

Comment:
Could be done once and the anonymous could be a simple pojo (package private).

Reply:
The idea was to not create another file/class just for that and to make it overridable. But I can also do the class.

Reply:
Your :+1:  is for creating the class or keeping the anonymous ?